drivers/filesystems/fastfat/misc.c

```diff
@@ -357,10 +402,41 @@ VfatQueueRequest(
     ASSERT(!(IrpContext->Flags & IRPCONTEXT_QUEUE) &&
            (IrpContext->Flags & IRPCONTEXT_COMPLETE));
 
+    Overflow = FALSE;
     IrpContext->Flags |= IRPCONTEXT_CANWAIT;
     IoMarkIrpPending(IrpContext->Irp);
-    ExInitializeWorkItem(&IrpContext->WorkQueueItem, VfatDoRequest, IrpContext);
-    ExQueueWorkItem(&IrpContext->WorkQueueItem, CriticalWorkQueue);
+
+    /* We should not block more than two worker threads per volume,
+     * or we might stop Cc from doing the work to unblock us.
+     * Add additional requests into the overflow queue instead and process
+     * them all in an existing worker thread (see VfatDoRequest above).
+     */
+    if (IrpContext->Stack->FileObject != NULL)
```

Comment:
Did you mean to check `DeviceObject` here, or is there a relation between FileObject and DeviceObject?

Reply:
Yeah I don't totally understand this part, but FileObject is right -- it's what fastfat_new does.